### ISSUE TITLE
Problem running PEX files with restricted permissions

### ISSUE THREAD

**Issue by jfinkhaeuser** [pantsbuild/pex#746]
When running PEX files, `PexInfo` determines the `install_cache` path based on the `pex_root` path, which in turn can be set in the `PEX-INFO` file - but isn't.

That means it defaults to `~/.pex` for writing the install cache - which fails when that path isn't writeable for the user. That can happen - is common practice, even - when securing a system against unwanted file system access. Writing to the home directory, for example, can modify environment variables in unwanted ways.

There's a number of things that could be improved in the approach of defaulting to `~/.pex`, but different requirements here would conflict with each other - so I'll leave that discussion to other people. The main point, though, is that it would be *very, very useful* if one could set the `pex_root` value that goes into the `PEX-INFO` file, so that you can effectively choose other paths for the `install_cache`. If additionally, `install_cache` could be overridden as well, that'd be excellent.

**Comment by jsirois** [comment]
> If additionally, install_cache could be overridden as well, that'd be excellent.

You should be able to do this. See `pex --help-variables`, but the `PEX_ROOT` environment variable should do what you want. Additionally, pex scans for a `.pexrc` file where you can define this variable as well. See:
https://github.com/pantsbuild/pex/blob/f5e00ad07a4943b101b4d17f94a1146856cfb585/pex/variables.py#L38-L59

Perhaps these two mechanisms solve your issue. Sealing pex root into the pex seems less than ideal since it's a machine specific setting and the pex is machine specific.

I'm going to close this as answered, but please report back with more questions or a success report.

**Comment by jfinkhaeuser** [comment]
Thanks, but this is precisely *not* what I need, going by the documentation that is. I do not care about setting the pex root during the build - during build, I have all the access to the home directory that I require. I need to control the install cache during execution, because that's when my environment is restricted.

I would like to urge you to re-open the issue.

**Comment by jfinkhaeuser** [comment]
I should add, I also tried it out just to confirm: it's not what I need.

**Comment by jfinkhaeuser** [comment]
To be clear: as a matter of fact, I do not need to set the pex root as such: all I need to be able to do is control the directory where the distribution gets unpacked to during execution. Unfortunately, that's tied to the pex root - I'm not currently sure why that would have to be, or should be the case.

I'd also be happy enough if the default was a globally writeable directory, such as `/tmp` - it's the `~/.pex` default that causes issues in restricted runtime environments.

**Comment by jfinkhaeuser** [comment]
To expand on this: assuming `/etc/pexrc` gets read during *execution* (which according to the code doesn't seem to be the case), it'd still be a global setting for all executables. That's not ideal, since each executable may have the `setuid` bit set, leading to an environment where a globally defined path is not accessible.

Related, many server programs that are `setuid` (or use a similar mechanism for changing the runtime user) deliberately restrict access to any paths that may lead to environment modifications, such as `~`.

What's needed for those is a neutral 'safe' location to write temporary files to - traditionally, this would be a `/var/cache` subdirectory for the install cache.

That's not portable, of course, but neither is `setuid` - so a pex mechanism for embedding such a safe path into the pex seems like a decent way to avoid excess complexity in pex, whilst also allowing to deal with such restricted situations.

**Comment by jsirois** [comment]
I think you may not have tested the pexrc approach or perhaps you were using older pex?

I find that the PEX_ROOT env var following:
```
~ $ mkdir -p /tmp/pex/issues/746
~ $ cd /tmp/pex/issues/746
# Prove contaminants are not present.
/tmp/pex/issues/746 $ cat /etc/pexrc
cat: /etc/pexrc: No such file or directory
/tmp/pex/issues/746 $ cat ~/.pexrc
cat: /home/jsirois/.pexrc: No such file or directory
/tmp/pex/issues/746 $ rm -rf ~/.pex && ls ~/.pex/
ls: cannot access '/home/jsirois/.pex/': No such file or directory
# Create an example pex.
/tmp/pex/issues/746 $ pex -c pex -o pex.pex pex==1.6.8
# Prove the PEX_ROOT env var is respected at runtime:
/tmp/pex/issues/746 $ PEX_ROOT=$PWD/pex_roots/via_explicit_env_var_at_runtime ./pex.pex --version
pex.pex 1.6.8
/tmp/pex/issues/746 $ ls -l ~/.pex/
ls: cannot access '/home/jsirois/.pex/': No such file or directory
/tmp/pex/issues/746 $ ls pex_roots/via_explicit_env_var_at_runtime
install
# Prove a sibling .pexrc is found and used - in this case to set PEX_ROOT implicitly.
/tmp/pex/issues/746 $ echo PEX_ROOT=$PWD/pex_roots/via_sibling_pexrc_at_runtime > .pexrc
/tmp/pex/issues/746 $ ./pex.pex --version
pex.pex 1.6.8
/tmp/pex/issues/746 $ ls -l ~/.pex/
ls: cannot access '/home/jsirois/.pex/': No such file or directory
/tmp/pex/issues/746 $ ls pex_roots/via_sibling_pexrc_at_runtime
install
```

**Comment by jsirois** [comment]
@jfinkhaeuser I've re-opened this issue since it has useful details and #748 does not. Please see my demonstration above of how PEX_ROOT is read at runtime and a sibling `.pexrc` can be deployed beside any pex to control its specific PEX_ROOT instead of using the global set in `/etc/pexrc`. Does this combination of features suit your needs?

**Comment by jfinkhaeuser** [comment]
So running through your steps that you posted does work. I'm not sure why it didn't seem to work before. I always got `~/.pex/` created again - it's possible that I previously used an older version, but how old would it have to have been? (Somewhat rhetorical question, unless the feature was very recently added).

So this is a step forward for me - but not the same as I need.

Imagine a suite of tools to be installed to e.g. `/usr/sbin`. Each of them is setuid some different system user (i.e. users without home directory, or home directory set to `/tmp` or some such). Actually, the setuid part isn't necessary for recreating the issue, it just illustrates that it's the users executing the binary that causes issues.

- With a global `/etc/pexrc`, the second invocation might run into issues when the first has created cache directories inaccessible to the second user.
- With a sibling `.pexrc` - well, there's only one sibling `.pexrc`, so it's as good as global.
- With a per-user or per-binary `PEX_ROOT` in the environment, the burden of making this possible is shifted to the user environment - which in the case of home-less system users is unpleasant to guarantee.

I'm not sure if this is sufficient information for why your approach is not what I'm looking for.

Another way of putting it is that a pex executable should just work like any other executable, and not rely on configuration to do so.

**Comment by jsirois** [comment]
Thanks, that backgroiund helps, specifically the fact you're installing >1 pex in in a flat dir like `/usr/sbin` and using setuid on top of that with different users. Agreed that setting PEX_ROOT from any of the possible sources pex respects at runtime does not work here. Say you have the folling pex binaries:
```
/usr/sbin/orchestrate
/usr/sbin/control
```

Normally you'd expect to be able to control them via something like `/etc/orchestrate` or `/etc/control`, but in this case, the user application, say `/etc/orchestrate`, has a generic underlying runtime - pex. So it seems to me there are only 2 viable paths:
1. have the option of embedding the PEX_ROOT setting into the pex at build time.
2. Enhance pexrc to support segregation of settings per binary.

You suggested 1 and I'm leery since this is a host-specific setting and pex generally has the goal of being host-independent mod being able to find a compatible python interpreter on the host at runtime.

The option in two could be done a few ways. Some ideas:
1. Support /etc/x.pexrc: Here x must be the basename of the pex binary.
2. Support /etc/pex.ini: Here the ini would contain a section per executable, say [x] following the example in 1 and [default] for a fallback.

I think 1 is most natural, it allows using `.x.pexrc` (sibling), `~/.x.pexrc` (home) and `/etc/x.pexrc` (global) for uniformity.

What do you think about this alternate approach? Does it or some variant work for you acceptably?

I'm not convinced this is the right way to move forward, but it does keep the pex non-host-specific and, fwict, allow full control on the target host via normalish setups / packaging standards.

Thanks for thinking through this with me.

**Comment by jfinkhaeuser** [comment]
Well, actually, what I'd prefer is for the cache to live outside of the `PEX_ROOT` in a place that's more accessible.

For each version of each dependency, there needs to be only one cached entry - it doesn't matter who creates it, as long as every user can read it.

That is, the parent directory that keeps the versioned entries needs to be world writable. In order not to make that an attack vector with injecting modified dependencies, it'd be good (if pex doesn't do that yet) for the pex created executable to contain a hash of the cache entry.

1. Write to the cache only if the entry does not exist.
1. If the entry exist and does not match the embedded hash, unpack it to a unique temporary location so that the binary runs. Sharing of cache data is then disabled.
1. If the entry exists and matches the embedded hash, yay, the cache is used and even shared across users.

If that lived in `/var/cache/pex` for example, then I think it'd be The Right Way (TM), too.

But I'm not sure that's a hole you want to dig.

**Comment by jsirois** [comment]
> Well, actually, what I'd prefer is for the cache to live outside of the PEX_ROOT in a place that's more accessible.

I don't understand this at all since `/etc/pexrc` lets you globally move `PEX_ROOT` to `/var/cache/pex` already.

> That is, the parent directory that keeps the versioned entries needs to be world writable. In order not to make that an attack vector with injecting modified dependencies, it'd be good (if pex doesn't do that yet) for the pex created executable to contain a hash of the cache entry.
...

Of the 3 items in your list:
1. Already implemented
2. If the entry doesn't exist (there is a hash mismatch), and it must be unpacked (is a non-zip-safe distribution) its is unpacked into the cache location as calculated from the current PEX_ROOT setting such that there is a cache hit on the next run in step 3. So, just like your #2 except instead of a unique temporary location, the cache is used.
3. This is done, but the contents of the cache location are trusted and not re-hashed at read time.

So, the missing piece in the recipe of creating a world writable `/var/cache/pex/{code,install}` dir and then setting up an `/etc/pexrc` with `PEX_PATH=/var/cache/pex` is checking the validity of cache hits by hashing their contents (fixing 3 above).

For example, today:
```
~ $ rm -rf /tmp/pex/issues/746 && mkdir -p /tmp/pex/issues/746 && cd /tmp/pex/issues/746
/tmp/pex/issues/746 $ pex -c pex -o pex.pex pex==1.6.8
/tmp/pex/issues/746 $ pex -c pex -o pex-requests.pex pex[requests]==1.6.8
/tmp/pex/issues/746 $ sudo rm -rf /var/cache/pex && sudo mkdir -p /var/cache/pex/install && sudo chmod 777 /var/cache/pex/install
/tmp/pex/issues/746 $ sudo bash -c 'echo PEX_ROOT=/var/cache/pex > /etc/pexrc'
/tmp/pex/issues/746 $ sudo -u nobody ./pex.pex --version
pex.pex 1.6.8
/tmp/pex/issues/746 $ ls -l /var/cache/pex/install
total 4
drwxr-xr-x 4 nobody nobody 4096 Aug  6 15:26 pex-1.6.8-py2.py3-none-any.whl.a8b41172679ac560bd75446e6a581a53f69c9bf4
/tmp/pex/issues/746 $ unzip -qc pex.pex PEX-INFO | jq .distributions
{
  "pex-1.6.8-py2.py3-none-any.whl": "a8b41172679ac560bd75446e6a581a53f69c9bf4"
}
/tmp/pex/issues/746 $ ./pex-requests.pex --version
pex-requests.pex 1.6.8
/tmp/pex/issues/746 $ ls -l /var/cache/pex/install/
total 24
drwxr-xr-x 4 jsirois jsirois 4096 Aug  6 15:29 certifi-2019.6.16-py2.py3-none-any.whl.e7502e9b88c49c666599cb7a4cc64504fe1f9022
drwxr-xr-x 4 jsirois jsirois 4096 Aug  6 15:29 chardet-3.0.4-py2.py3-none-any.whl.c4332e34e38b781695ce775973fe40663558a897
drwxr-xr-x 4 jsirois jsirois 4096 Aug  6 15:29 idna-2.8-py2.py3-none-any.whl.645e82ab4157606d15a8766060dd690deebcc4f2
drwxr-xr-x 4 nobody  nobody  4096 Aug  6 15:26 pex-1.6.8-py2.py3-none-any.whl.a8b41172679ac560bd75446e6a581a53f69c9bf4
drwxr-xr-x 4 jsirois jsirois 4096 Aug  6 15:29 requests-2.22.0-py2.py3-none-any.whl.bb3699cc576016729d42f9bf4eccfdbcb292ec50
drwxr-xr-x 4 jsirois jsirois 4096 Aug  6 15:29 urllib3-1.25.3-py2.py3-none-any.whl.cfa3106651fa0507d9a96160ed69d5cf13c5623e
/tmp/pex/issues/746 $ unzip -qc pex-requests.pex PEX-INFO | jq .distributions
{
  "certifi-2019.6.16-py2.py3-none-any.whl": "e7502e9b88c49c666599cb7a4cc64504fe1f9022",
  "chardet-3.0.4-py2.py3-none-any.whl": "c4332e34e38b781695ce775973fe40663558a897",
  "idna-2.8-py2.py3-none-any.whl": "645e82ab4157606d15a8766060dd690deebcc4f2",
  "pex-1.6.8-py2.py3-none-any.whl": "a8b41172679ac560bd75446e6a581a53f69c9bf4",
  "requests-2.22.0-py2.py3-none-any.whl": "bb3699cc576016729d42f9bf4eccfdbcb292ec50",
  "urllib3-1.25.3-py2.py3-none-any.whl": "cfa3106651fa0507d9a96160ed69d5cf13c5623e"
}
/tmp/pex/issues/746 $ echo 'print(">>> BOOM"); import sys; sys.exit(42)' >> /var/cache/pex/install/requests-2.22.0-py2.py3-none-any.whl.bb3699cc576016729d42f9bf4eccfdbcb292ec50/requests-2.22.0-py2.py3-none-any.whl/requests/__init__.py
/tmp/pex/issues/746 $ tail -2 /var/cache/pex/install/requests-2.22.0-py2.py3-none-any.whl.bb3699cc576016729d42f9bf4eccfdbcb292ec50/requests-2.22.0-py2.py3-none-any.whl/requests/__init__.py
warnings.simplefilter('default', FileModeWarning, append=True)
print(">>> BOOM"); import sys; sys.exit(42)
/tmp/pex/issues/746 $ ./pex-requests.pex isort
>>> BOOM
/tmp/pex/issues/746 $ echo $?
42
```

So, adding a flag to turn on cache validation at pex runtime seems reasonable to me. It's really a general security hole that exists today and should be fixed anyhow if not on by default. So a very small amount of code here:
https://github.com/pantsbuild/pex/blob/f5e00ad07a4943b101b4d17f94a1146856cfb585/pex/environment.py#L180-L187

Basically:
```python
if os.path.exists(cached_location) and dist_digest = CacheHelper.dir_hash(cached_location):
  ...
```

Where `CacheHelper.dir_hash` would be a new function that worked much like the existing `CacheHelper.zip_hash`.

Does adding a build-time flag + new PexInfo option + implementing the bit of code mentioned above seem reasonable to you? I think it solves your problem and would be a generally useful feature anyhow for untrusted environments.

**Comment by jfinkhaeuser** [comment]
> > Well, actually, what I'd prefer is for the cache to live outside of the PEX_ROOT in a place that's more accessible.
> 
> I don't understand this at all since `/etc/pexrc` lets you globally move `PEX_ROOT` to `/var/cache/pex` already.

`PEX_ROOT`, by my reading, may contain more than just the cache. It's the cache that may need to be shared between different users. That's why I'd move the cache outside of `PEX_ROOT`, because it may warrant special treatment. If you disagree, you know the code better than me.

>     1. Already implemented
Sans the world writable, but yes.
 
>     2. If the entry doesn't exist (there is a hash mismatch), and it must be unpacked (is a non-zip-safe distribution) its is unpacked into the cache location as calculated from the current PEX_ROOT setting such that there is a cache hit on the next run in step 3. So, just like your #2 except instead of a unique temporary location, the cache is used.

Well, the temporary location is really a fix for this issue:

```bash
# Create test directory, owned by root. Make parent world writable
$ sudo mkdir -p /tmp/test/bar
$ sudo chmod 0777 /tmp/test/

# Create an entry in the parent, and one in the root-owned test directory
$ sudo touch /tmp/test/foo
$ sudo touch /tmp/test/bar/baz

# Try to remove immediate child - works!
$ rm -rf /tmp/test/foo 

# Try to remove root-owned directory
$ rm -rf /tmp/test/bar/
rm: cannot remove '/tmp/test/bar/baz': Permission denied
```

(Where sudo is just a stand-in for any other user). Basically, a world-writable cache root does not mean that you can remove individual entries created by other users. So *reading* entries created by other users is likely (!?) sensible; most system's default permissions allow for that. *Modifying* entries created by other users requires setting permissions after creating the cache entries - which is of course a possibility. I was going for another route, where you never try to do that.

If the cache entry paths are themselves versioned, then creating two cache entries for different versions of the same package under different users is going to be fine. There is no "upgrade problem". The only problem exists when a cache entry doesn't match the embedded hash.

Now there are two diverging philosophies here: one says that a corrupt cache is enough of an issue that one should fail loudly. The other, which is the philosophy I'm following here in this issue, is that system tools should never fail if they can recover (hence the `/usr/sbin` location in my example). The safest way to never fail is to just not use the cache when it's corrupt, hence the temporary location.

A different approach might be to use the entry hash as the path, so you can't have mismatching versions - that's ok if the pex tries to find an entry in the cache (use the hash as the path, if it exist it's fine), but doesn't solve the problem when the cache entry actually hashes to something else than it's path. Then you're back to not being sure whether you can modify the entry.

(As an aside, an issue with a shared cache is that two executables running simultaneously might find partially populated cache entries - you'd either have to create cache entries atomically via a directory move, or recover from that by treating it just like a corrupt cache and not use the cache in the second invocation - a third invocation might find the completed cache entry and use it again.)

I hope this makes sense. Whether it's the solution you prefer is another matter.

>     3. This is done, but the contents of the cache location are trusted and not re-hashed at read time.
> 
> 
> So, the missing piece in the recipe of creating a world writable `/var/cache/pex/{code,install}` dir and then setting up an `/etc/pexrc` with `PEX_PATH=/var/cache/pex` is checking the validity of cache hits by hashing their contents (fixing 3 above).

Again repeating the system tool perspective, it'd be great if this was actually the default behaviour rather than something that needs to be configured. That doesn't mean configuring it to some other location wouldn't work - but I'd prefer executables to just work, with or without configuration.

I'm certain what you're proposing brings me a huge step closer, but if I can convince you of the extra effort I'm proposing, then that'd be even cooler.

FWIW, this is why I suggested embedding the cache directory into the pex - this way every pex file can pick its own, and you can circumvent all this permission management. But I do agree that default behaviour that always works would be preferable.

**Comment by jfinkhaeuser** [comment]
Or, in pseudo-Python:

```python
if not cache_entry.exists:
  create_global_cache_entry()
  use_global_cache_entry()
else:
  if cache_entry.matches(cache_entry_hash):
    use_global_cache_entry()
  else:
    create_temporary_cache_entry()
    use_temporary_cache_entry()
```

**Comment by jsirois** [comment]
@jfinkhaeuser thanks for working through the problem space with me. This definitely softened me to the idea. We now have `--runtime-pex-root` for setting the PEX_ROOT to use in the PEX metadata from #780 and #929 adds fallback to a writeable temporary directory with a warning to stderr if it so happens the pex is deployed somewhere where its `--runtime-pex-root` does not exist or cannot be written to. These features will ship with Pex 2.1.7 shortly.

**Comment by jfinkhaeuser** [comment]
That's excellent news to me! I will try this out, but am swamped this week - maybe next? Either way, I can let you know if it works for me!

**Comment by benjyw** [comment]
@jfinkhaeuser Hi, hope you're well. Just wanted to check that this solved the problem for you? Thanks.

**Comment by jfinkhaeuser** [comment]
Ha! Thanks for asking. I'd completely forgotten about this due to all the other things going on for me. I'm sorry, I dropped the ball there. I will check it out today.

**Comment by benjyw** [comment]
No rush - just wanted to make sure we didn't close this prematurely.

**Comment by jfinkhaeuser** [comment]
It works beautifully, thank you so much!

**Comment by benjyw** [comment]
Glad to hear it!
